File: plugins/plugin-minecraft/rust/src/providers/waypoints.rs
use serde_json::json;
use std::collections::HashMap;

use super::ProviderResult;
use crate::services::WaypointsService;

/// Provider that returns all saved waypoints
pub fn get_waypoints(waypoints_service: &WaypointsService) -> ProviderResult {
    let wp_list = waypoints_service.list_waypoints();

    if wp_list.is_empty() {
        return ProviderResult {
            text: "No waypoints saved.".to_string(),
            values: {
                let mut v = HashMap::new();
                v.insert("count".to_string(), json!(0));
                v
            },
            data: {
                let mut d = HashMap::new();
                d.insert("waypoints".to_string(), json!([]));
                d
            },
        };
    }

    let lines: Vec<String> = wp_list
        .iter()
        .map(|w| format!("- {}: ({:.1}, {:.1}, {:.1})", w.name, w.x, w.y, w.z))
        .collect();

    let waypoints_data: Vec<serde_json::Value> = wp_list
        .iter()
        .map(|w| {
            json!({
                "name": w.name,
                "x": w.x,
                "y": w.y,
                "z": w.z,
                "createdAt": w.created_at.to_rfc3339(),
            })
        })
        .collect();

    ProviderResult {
        text: format!("Waypoints:\n{}", lines.join("\n")),
        values: {
            let mut v = HashMap::new();
            v.insert("count".to_string(), json!(wp_list.len()));
            v
        },
        data: {
            let mut d = HashMap::new();
            d.insert("waypoints".to_string(), json!(waypoints_data));
            d
        },
    }
}

File: plugins/plugin-minecraft/typescript/src/providers/waypoints.ts
import type {
  IAgentRuntime,
  Memory,
  Provider,
  ProviderResult,
  State,
} from "@elizaos/core";
import {
  WAYPOINTS_SERVICE_TYPE,
  type WaypointsService,
} from "../services/waypoints-service.js";

export const minecraftWaypointsProvider: Provider = {
  name: "MC_WAYPOINTS",
  description: "Saved Minecraft waypoints (names and coordinates)",
  get: async (
    runtime: IAgentRuntime,
    _message: Memory,
    _state?: State,
  ): Promise<ProviderResult> => {
    const service = runtime.getService<WaypointsService>(
      WAYPOINTS_SERVICE_TYPE,
    );
    if (!service) {
      return {
        text: "Waypoints service not available",
        values: { count: 0 },
        data: { waypoints: [] },
      };
    }

    const list = service.listWaypoints();
    const lines = list.map(
      (w) =>
        `- ${w.name}: (${w.x.toFixed(1)}, ${w.y.toFixed(1)}, ${w.z.toFixed(1)})`,
    );
    return {
      text: list.length
        ? `Waypoints:\n${lines.join("\n")}`
        : "No waypoints saved.",
      values: { count: list.length },
      data: {
        waypoints: list.map((w) => ({
          name: w.name,
          x: w.x,
          y: w.y,
          z: w.z,
          createdAt: w.createdAt.toISOString(),
        })),
      },
    };
  },
};
